### ISSUE TITLE
need kubelet executable but none of the candidates are running

### ISSUE THREAD

**Issue by SakshiSaini17092** [aquasecurity/kube-bench#328]
I am trying to set-up kube-bench on openshift-3.11 version.

I have cloned the latest git repo and ran this command 
`oc apply -f job.yaml`

but when I ran this command `oc get pod`
<pre>
NAME                                READY     STATUS      RESTARTS   AGE
docker-registry-84cd6bfd4-l95ck     1/1       Running     0          1d
docker-registry-84cd6bfd4-ncfwj     1/1       Running     0          1d
kube-bench-hrlnf                    0/1       Error       0          36s
kube-bench-nnjtz                    0/1       Error       0          33s
kube-bench-rcbdf                    0/1       Error       0          3s
kube-bench-rftjp                    0/1       Error       0          23s
kube-hunter-gc6s5                   0/1       Completed   0          9h
registry-console-7b9b55967c-tfm7b   1/1       Running     0          1d
router-55844474cd-7fpxm             1/1       Running     0          1d
router-55844474cd-bthfz             1/1       Running     0          1d
</pre>

this is the output .
the logs of the kube-bench pods are `need kubelet executable but none of the candidates are running`

I have also tried by specifying the master version as ocp-3.11 in job.yaml file.
but this time the error is different 
`need apiserver executable but none of the candidates are running`

I have tried by changing the master version to ocp-3.11 to wherever it's mentioned. but not able to debug it can you please help me.

this is the output of `oc get nodes`
<pre>
NAME            STATUS    ROLES           AGE       VERSION
10.148.35.195   Ready     compute,infra   1d        v1.11.0+d4cacc0
10.148.35.205   Ready     compute,infra   1d        v1.11.0+d4cacc0
</pre>

I am using kubectl version
<pre>
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-20T04:49:16Z", GoVersion:"go1.12.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0+d4cacc0", GitCommit:"d4cacc0", GitTreeState:"clean", BuildDate:"2019-04-10T17:49:11Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
</pre>

**Comment by awalker125** [comment]
We might be seeing something similar

```
[centos@xxxxxxx]$ kubectl logs kube-bench-master-zk29p

need controllermanager executable but none of the candidates are running
```

I think there is new latest version push a few hours ago. I wonder if it has caused it?

**Comment by ttousai** [comment]
@SakshiSaini17092 @awalker125 some more info will be helpful, please do the following and share the info:
```
- Please spend a small amount of time giving due diligence to the issue tracker. Your issue might be a duplicate.
- Note the version of kube-bench you are running and the command line options you are using.
- Note the version of kubernetes you are running (from `kubectl version` or `oc version` for Openshift).
- Set `-v 10` command line option and save the log output. Please paste this into your issue.
  Make a copy of the job.yaml file and change the command field to ['kube-bench', '-v10'] command field
```

**Comment by harishspqr** [comment]
Looks like something is messed up with the latest image push because it was working yesterday. In my case its not able to find kube-controller-manager when its clearly running. 

I did added some kubeadm extraargs to fix the issues reported by kube-bench during this run though. 

```kubectl logs kube-bench-master-gs89w
I0627 19:14:12.294489   25495 util.go:122] Looking for config for version 1.13
I0627 19:14:12.294536   25495 util.go:127] Looking for config file: cfg/1.13/master.yaml
I0627 19:14:12.294759   25495 common.go:228] Using config file: cfg/1.13/config.yaml
I0627 19:14:12.294834   25495 common.go:74] Using benchmark file: cfg/1.13/master.yaml
I0627 19:14:12.301688   25495 util.go:102] Component apiserver uses running binary kube-apiserver
I0627 19:14:12.310063   25495 util.go:102] Component scheduler uses running binary kube-scheduler
I0627 19:14:12.322770   25495 util.go:43] [ps -C kube-controller-manager -o cmd --no-headers]: exit status 1
I0627 19:14:12.322850   25495 util.go:305] executable 'kube-controller-manager' not running
I0627 19:14:12.328926   25495 util.go:43] [ps -C hyperkube -o cmd --no-headers]: exit status 1
I0627 19:14:12.328968   25495 util.go:305] executable 'hyperkube controller-manager' not running
I0627 19:14:12.334774   25495 util.go:43] [ps -C hyperkube -o cmd --no-headers]: exit status 1
I0627 19:14:12.334816   25495 util.go:305] executable 'hyperkube kube-controller-manager' not running
I0627 19:14:12.340474   25495 util.go:43] [ps -C controller-manager -o cmd --no-headers]: exit status 1
I0627 19:14:12.340515   25495 util.go:305] executable 'controller-manager' not running```

I can verify that kube-controller-manager is running on controlplane. 
```ps -C kube-controller-manager -o cmd --no-headers
kube-controller-manager --cloud-provider=aws --profiling=false --address=127.0.0.1 --allocate-node-cidrs=true --authentication-kubeconfig=/etc/kubernetes/controller-manager.conf -```

**Comment by lizrice** [comment]
@harishspqr your log output is interesting because it shows it's finding kube-apiserver and kube-scheduler just fine, but then failing to find kube-controller-manager. Please could you confirm, when you checked manually with the `ps -C kube-controller-manager` command was that on the same node as where the logs came from?

**Comment by SakshiSaini17092** [comment]
@ttousai I am using oc version:
<pre>
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0

openshift v3.11.104
kubernetes v1.11.0+d4cacc0

</pre>



and I am using this command to run kube-bench
    `oc apply -f job.yaml`

and in the job.yaml file in line number 15 I have changed the command to this:

`command: ["kube-bench", "master" , "--version", "ocp-3.11" , "node","--version","1.11"]` as mentioned in closed issues.

but this gives error `need apiserver executable but none of the candidates are running`

the output of running this job as `oc apply -f job.yaml -v 10`

[output.txt](https://github.com/aquasecurity/kube-bench/files/3338658/output.txt)

**Comment by pradeepkondaveti** [comment]
My issue is same as @awalker125 , latest image is working fine for Node but failing for master with the following error. 

`need controllermanager executable but none of the candidates are running`

I can see the next old tag was `ocp-configs` available from docker hub which is 5 month old.

**Comment by lizrice** [comment]
I can reproduce the behaviour for failing to find the controllermanager. This works fine when running the executable directly on the host, but it's failing when built into a container. Investigating.

**Comment by lizrice** [comment]
This is essentially fixed by #326. Here's what's happening:

```bash
$ docker run --pid=host -it alpine:3.10 sh
/ # apk --no-cache add procps
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
(1/6) Installing libintl (0.19.8.1-r4)
(2/6) Installing ncurses-terminfo-base (6.1_p20190518-r0)
(3/6) Installing ncurses-terminfo (6.1_p20190518-r0)
(4/6) Installing ncurses-libs (6.1_p20190518-r0)
(5/6) Installing libproc (3.3.15-r0)
(6/6) Installing procps (3.3.15-r0)
Executing busybox-1.30.1-r2.trigger
OK: 14 MiB in 20 packages
/ # ps -C kube-controller
  PID TTY          TIME CMD
 6988 ?        00:01:28 kube-controller
/ # ps -C kube-controller-manager
  PID TTY          TIME CMD
```
Note there is no output when we look for the full name, which is over 15 chars long.

**Comment by Slach** [comment]
i have a same error message i just run
```bash
KUBECONFIG=/etc/kubernetes/admin.conf kubectl apply -f https://raw.githubusercontent.com/aquasecurity/kube-bench/master/job.yaml
```
on my kubernetes 1.15 + containerd.io as CRI

**Comment by SakshiSaini17092** [comment]
> i have a same error message i just run
> 
> ```shell
> KUBECONFIG=/etc/kubernetes/admin.conf kubectl apply -f https://raw.githubusercontent.com/aquasecurity/kube-bench/master/job-master.yaml
> ```
> 
> on my kubernetes 1.15 + containerd.io as CRI
> 
> ```
> root@k8s-master1:/home/vagrant# crictl ps -a
> CONTAINER ID        IMAGE               CREATED             STATE               NAME                      ATTEMPT             POD ID
> 8fb3da7231617       0a1b4e8c92d2b       1 second ago        Exited              kube-bench                0                   c93d073c8b9a8
> ```
> 
> after that in logs i see
> 
> ```
> root@k8s-master1:/home/vagrant# crictl logs 8fb3da7231617
> need controllermanager executable but none of the candidates are running
> ```
> 
> controller-manager running and visible on host os
> 
> ```
> root@k8s-master1:/home/vagrant# ps -ef | grep controller | grep -v grep
> root      3181  3131  1 19:08 ?        00:00:09 kube-controller-manager --allocate-node-cidrs=true --authentication-kubeconfig=/etc/kubernetes/controller-manager.conf --authorization-kubeconfig=/etc/kubernetes/controller-manager.conf --bind-address=127.0.0.1 --client-ca-file=/etc/kubernetes/pki/ca.crt --cluster-cidr=10.244.0.0/16 --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt --cluster-signing-key-file=/etc/kubernetes/pki/ca.key --controllers=*,bootstrapsigner,tokencleaner --feature-gates=RotateKubeletServerCertificate=true --kubeconfig=/etc/kubernetes/controller-manager.conf --leader-elect=true --node-cidr-mask-size=24 --profiling=false --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --root-ca-file=/etc/kubernetes/pki/ca.crt --service-account-private-key-file=/etc/kubernetes/pki/sa.key --service-cluster-ip-range=10.99.0.0/16 --terminated-pod-gc-threshold=100 --use-service-account-credentials=true
> ```



> i have a same error message i just run
> 
> ```shell
> KUBECONFIG=/etc/kubernetes/admin.conf kubectl apply -f https://raw.githubusercontent.com/aquasecurity/kube-bench/master/job-master.yaml
> ```
> 
> on my kubernetes 1.15 + containerd.io as CRI
> 
> ```
> root@k8s-master1:/home/vagrant# crictl ps -a
> CONTAINER ID        IMAGE               CREATED             STATE               NAME                      ATTEMPT             POD ID
> 8fb3da7231617       0a1b4e8c92d2b       1 second ago        Exited              kube-bench                0                   c93d073c8b9a8
> ```
> 
> after that in logs i see
> 
> ```
> root@k8s-master1:/home/vagrant# crictl logs 8fb3da7231617
> need controllermanager executable but none of the candidates are running
> ```
> 
> controller-manager running and visible on host os
> 
> ```
> root@k8s-master1:/home/vagrant# ps -ef | grep controller | grep -v grep
> root      3181  3131  1 19:08 ?        00:00:09 kube-controller-manager --allocate-node-cidrs=true --authentication-kubeconfig=/etc/kubernetes/controller-manager.conf --authorization-kubeconfig=/etc/kubernetes/controller-manager.conf --bind-address=127.0.0.1 --client-ca-file=/etc/kubernetes/pki/ca.crt --cluster-cidr=10.244.0.0/16 --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt --cluster-signing-key-file=/etc/kubernetes/pki/ca.key --controllers=*,bootstrapsigner,tokencleaner --feature-gates=RotateKubeletServerCertificate=true --kubeconfig=/etc/kubernetes/controller-manager.conf --leader-elect=true --node-cidr-mask-size=24 --profiling=false --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --root-ca-file=/etc/kubernetes/pki/ca.crt --service-account-private-key-file=/etc/kubernetes/pki/sa.key --service-cluster-ip-range=10.99.0.0/16 --terminated-pod-gc-threshold=100 --use-service-account-credentials=true
> ```

Hey can you please tell me which command you used to SSH into node. I mean I have searched and tried a lot but not able to do. I only find how to do rsh on a pod but that was not helpful. I want to check the path of kubectl on my node.

**Comment by Slach** [comment]
> Hey can you please tell me which command you used to SSH into node. I mean I have searched and tried a lot but not able to do. I only find how to do rsh on a pod but that was not helpful. I want to check the path of kubectl on my node.

i just use local development environment with vagrant
and use ```vagrant ssh k8s-master1``` to connect over SSH inside HOST OS and it's not ssh mode inside pod
 
and kubectl inside my k8s-vagrant in /usr/bin/kubelet

**Comment by lizrice** [comment]
@slach we had some issues with our automated builds and I suspect the kube-bench version you pulled didn't have the fix for #326 yet. Would you mind trying again to see if you still see the same issue?

**Comment by SakshiSaini17092** [comment]
Thank you everyone for your help!.

I found the solution (need to rebuild the DockerImage with changed configurations files ) and I am closing this issue now.
